File: gateway/gateway_stack.py
# Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
# SPDX-License-Identifier: MIT-0

"""
CDK Stack for AgentCore Gateway IAM Role
"""

import aws_cdk as cdk
from aws_cdk import (
    Stack,
    aws_iam as iam,
    aws_ssm as ssm,
)
from constructs import Construct


class AgentCoreGatewayStack(Stack):
    """CDK Stack that creates the IAM role for AgentCore Gateway."""

    def __init__(self, scope: Construct, construct_id: str, **kwargs) -> None:
        super().__init__(scope, construct_id, **kwargs)

        # Create the IAM role for AgentCore Gateway
        self.gateway_role = self._create_gateway_role()

        # Store the role ARN in SSM Parameter Store
        self._store_role_arn()

        # Output the role ARN
        cdk.CfnOutput(
            self,
            "GatewayRoleArn",
            value=self.gateway_role.role_arn,
            description="ARN of the AgentCore Gateway IAM role",
            export_name=f"AgentCoreGatewayRoleArn",
        )

    def _create_gateway_role(self) -> iam.Role:
        """Create the IAM role for AgentCore Gateway."""

        # Retrieve the Lambda function ARN from SSM Parameter Store
        lambda_function_arn = ssm.StringParameter.value_for_string_parameter(
            self, "/sagemaker-async/lambda-function-arn"
        )

        # Create the role with basic service principal first
        role = iam.Role(
            self,
            "AgentCoreGatewayRole",
            role_name=f"agentcore-gateway-role",
            assumed_by=iam.ServicePrincipal("bedrock-agentcore.amazonaws.com"),
            description=f"IAM role for AgentCore Gateway",
        )

        # Add the conditions to the assume role policy
        role.assume_role_policy.add_statements(
            iam.PolicyStatement(
                sid="AssumeRolePolicyWithConditions",
                effect=iam.Effect.ALLOW,
                principals=[iam.ServicePrincipal("bedrock-agentcore.amazonaws.com")],
                actions=["sts:AssumeRole"],
                conditions={
                    "StringEquals": {"aws:SourceAccount": cdk.Aws.ACCOUNT_ID},
                    "ArnLike": {
                        "aws:SourceArn": f"arn:aws:bedrock-agentcore:{cdk.Aws.REGION}:{cdk.Aws.ACCOUNT_ID}:*"
                    },
                },
            )
        )

        # Add permissions policy with specific Lambda function ARN
        role.add_to_policy(
            iam.PolicyStatement(
                sid="LambdaInvokePermissions",
                effect=iam.Effect.ALLOW,
                actions=[
                    "lambda:InvokeFunction",
                ],
                resources=[lambda_function_arn],
            )
        )

        # Add tags
        cdk.Tags.of(role).add("Service", "bedrock-agentcore")
        cdk.Tags.of(role).add("Component", "gateway")

        return role

    def _store_role_arn(self) -> None:
        """Store the role ARN in SSM Parameter Store."""
        ssm.StringParameter(
            self,
            "GatewayRoleArnParameter",
            parameter_name=f"/agentcore-gateway/role-arn",
            string_value=self.gateway_role.role_arn,
            #
            description=f"ARN of the AgentCore Gateway IAM role",
            tier=ssm.ParameterTier.STANDARD,
        )

File: cdk/lib/gateway-stack.ts
import * as cdk from 'aws-cdk-lib';
import * as iam from 'aws-cdk-lib/aws-iam';
import * as bedrockagentcore from 'aws-cdk-lib/aws-bedrockagentcore';
import { Construct } from 'constructs';

export interface GatewayStackProps extends cdk.StackProps {
  /**
   * Project name used for resource naming and tagging
   * @default 'protein-engineering'
   */
  readonly projectName?: string;

  /**
   * Name for the AgentCore Gateway
   * @default 'agentcore-gateway'
   */
  readonly gatewayName?: string;

  /**
   * Description for the AgentCore Gateway
   * @default 'AgentCore Gateway for protein engineering agent with MCP integration'
   */
  readonly gatewayDescription?: string;
}
export class GatewayStack extends cdk.Stack {
  public readonly gateway: bedrockagentcore.CfnGateway;
  public readonly gatewayTarget: bedrockagentcore.CfnGatewayTarget;
  public readonly gatewayRole: iam.Role;

  constructor(scope: Construct, id: string, props?: GatewayStackProps) {
    super(scope, id, props);

    // Extract configuration with defaults
    const projectName = props?.projectName ?? 'protein-engineering';
    const gatewayName = props?.gatewayName ?? 'agentcore-gateway';
    const gatewayDescription = props?.gatewayDescription ?? 
      'AgentCore Gateway for protein engineering agent with MCP integration';

    this.gatewayRole = new iam.Role(this, 'GatewayRole', {
      roleName: `${projectName}-agentcore-gateway-role`,
      description: 'Execution role for Amazon Bedrock AgentCore Gateway to invoke Lambda functions',
      assumedBy: new iam.ServicePrincipal('bedrock-agentcore.amazonaws.com', {
        conditions: {
          StringEquals: {
            'aws:SourceAccount': this.account,
          },
          ArnLike: {
            'aws:SourceArn': `arn:${this.partition}:bedrock-agentcore:${this.region}:${this.account}:*`,
          },
        },
      }),
    });

    this.gatewayRole.addToPolicy(
      new iam.PolicyStatement({
        effect: iam.Effect.ALLOW,
        actions: ['lambda:InvokeFunction'],
        resources: [lambdaFunctionArn],
      })
    );

    this.gateway = new bedrockagentcore.CfnGateway(this, 'AgentCoreGateway', {
      name: gatewayName,
      description: gatewayDescription,
      roleArn: this.gatewayRole.roleArn,
      protocolType: 'MCP',
      authorizerType: 'CUSTOM_JWT',
      authorizerConfiguration: {
        customJwtAuthorizer: {
          discoveryUrl: cognitoDiscoveryUrl,
          allowedClients: [cognitoClientId],
        },
      },
      protocolConfiguration: {
        mcp: {
          searchType: 'SEMANTIC',
        },
      },
      tags: {
        Project: projectName,
        Component: 'Gateway',
        ManagedBy: 'CDK',
      },
    });

    this.gatewayTarget = new bedrockagentcore.CfnGatewayTarget(this, 'GatewayTarget', {
      name: 'protein-engineering-lambda',
      description: 'Lambda target for protein variant effect prediction',
      gatewayIdentifier: this.gateway.ref,
      targetConfiguration: {
        mcp: {
          lambda: {
            lambdaArn: lambdaFunctionArn,
            toolSchema: {
              inlinePayload: [
                {
                  name: 'invoke_endpoint',
                  description: 'Submit a job to predict the variant effects for an amino acid sequence',
                  inputSchema: {
                    type: 'object',
                    properties: {
                      sequence: {
                        type: 'string',
                        description: 'Amino acid sequence',
                      },
                    },
                    required: ['sequence'],
                  },
                },
                {
                  name: 'get_results',
                  description: 'Check if a variant effect prediction job has completed and retrieve results',
                  inputSchema: {
                    type: 'object',
                    properties: {
                      output_id: {
                        type: 'string',
                        description: 'output_id returned by invoke_endpoint',
                      },
                    },
                    required: ['output_id'],
                  },
                },
              ],
            },
          },
        },
      },
      credentialProviderConfigurations: [
        {
          credentialProviderType: 'GATEWAY_IAM_ROLE',
        },
      ],
    });
